walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,64 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    ArrayList<TimeRange> validTimes = new ArrayList<TimeRange>();
+    Collection<String> wantedAttendees = request.getAttendees();
+    ArrayList<TimeRange> occupiedTimes = new ArrayList<TimeRange>();
+
+    TimeRange whole_day = TimeRange.WHOLE_DAY;
+
+    if(wantedAttendees.size() == 0) {
+        validTimes.add(whole_day);
+        return validTimes;
+    }
+
+    if(request.getDuration() > whole_day.duration()) {
+        return validTimes;
+    }
+
+    for(Event schedule : events) {
+        Collection<String> busyAttendees = schedule.getAttendees();
+
+        for(String person:wantedAttendees) {
+            if(busyAttendees.contains(person)) {
+                occupiedTimes.add(schedule.getWhen());
+            }
+        }
+    }
+
+    if(occupiedTimes.size()==0) {
+        validTimes.add(whole_day);
+        return validTimes;
+    }
+
+    Collections.sort(occupiedTimes, TimeRange.ORDER_BY_START);
+
+    TimeRange marker = TimeRange.START;
+
+    for(TimeRange meeting : occupiedTimes){
+        if(meeting.start() > marker.end() && meeting.start()-marker.end() >= request.getDuration()) {
+            validTimes.add(TimeRange.fromStartEnd(marker.end(), meeting.start()-1, true));
+        }
+        if(marker.end() < meeting.end()) {
+            marker = meeting;
+        }
+    }
+    if(marker.end()!=whole_day.end() && whole_day.end()-marker.end()>=request.getDuration()){
```

Comment:
Might want to add a comment above this if statement, explaining to the reader why we need to do this last check outside of the loop

Reply:
Should I be putting in more comments for the other parts of my code?

Reply:
No, you should be fine! Commenting is an interesting thing. Typically, for in line comments like this, we only want to do it if it adds something extra that's not obvious from reading the code. Good variable naming and good structure typically mean fewer comments. But sometimes we need to do something and it isn't instantly obvious as to why (or maybe a line of code is very complicated to read). In these cases, having some plain english to describe the code is helpful.